src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp

```diff
@@ -533,7 +554,24 @@ std::unique_ptr<IObjectStorage> S3ObjectStorage::cloneObjectStorage(
     return std::make_unique<S3ObjectStorage>(
         std::move(new_client), std::move(new_s3_settings),
         version_id, s3_capabilities, new_namespace,
-        endpoint, object_key_prefix);
+        endpoint, object_key_prefix, disk_name);
+}
+
+
+BlobStorageLogWriter S3ObjectStorage::getBlobStorageLog()
+{
+#ifndef CLICKHOUSE_KEEPER_STANDALONE_BUILD /// Keeper standalone build doesn't have context
+    /// Make a copy with local properties like query_id, object path, etc
+    BlobStorageLogWriter blob_storage_log(Context::getGlobalContextInstance()->getBlobStorageLog());
+    blob_storage_log.disk_name = disk_name;
+
```

Comment:
May be it could be better if `getBlobStorageLog` returned a pointer instead of an object and it will be nullptr if `context->getBlobStorageLog()` returns nullptr (or if it is disabled with query setting), this will allow to avoid redundant work with checking query context below and then calling `addEvent` if the no event will actually be added because `log` is nullptr. Also there are places where we do `addEvent` in a loop for many objects, but there is no point in doing it if log is not initialized or disabled.

Reply:
Replaced `BlobStorageLogWriter` with `BlobStorageLogWriterPtr`. Initially I didn't want to check

```cpp
if (blob_log)
	blob_log->addEvent(...)
```

each time, but as you mentioned having plain object `BlobStorageLogWriter` also have some drawbacks, so pointer is good.